### PR TITLE
enhancement(vrl): Added case sensitive option to enrichment VRL functions

### PR DESCRIPTION
**Pull request by StephenWakely**
ref #8815 

This adds a case sensitive option (defaulting to true) to the `get_enrichment_table_record` and `find_enrichment_table_records` VRL functions.

I'll add a followup PR to get the docs in shortly!

### CODE REVIEW

File: lib/enrichment/src/find_enrichment_table_records.rs [hunk 1]
```diff
@@ -93,13 +114,18 @@ impl Expression for FindEnrichmentTableRecordsFn {
...
+            self.case_sensitive,
+            &self.condition,
+        )?);
         Ok(())
     }
 
     fn type_def(&self, _: &state::Compiler) -> TypeDef {
         TypeDef::new()
-            .fallible()
+            .infallible()
```
Comment (jszwedko) [inline lib/enrichment/src/find_enrichment_table_records.rs:128]: Is this change related? Or just something you noticed?

**Comment by lucperkins** [comment]
@StephenWakely *cough* docs *cough*

File: src/enrichment_tables/file.rs [hunk 2]
```diff
@@ -178,22 +179,20 @@ impl File {
...
+                Some(idx) => match (case, &row[idx], value) {
+                    (Case::Insensitive, Value::Bytes(bytes1), Value::Bytes(bytes2)) => {
                         match (std::str::from_utf8(bytes1), std::str::from_utf8(bytes2)) {
                             (Ok(s1), Ok(s2)) => s1.to_lowercase() == s2.to_lowercase(),
                             (Err(_), Err(_)) => bytes1 == bytes2,
                             _ => false,
                         }
                     }
-                    (value1, value2) => value1 == value2,
+                    (_, value1, value2) => value1 == value2,
```
Comment (jszwedko) [inline src/enrichment_tables/file.rs:195]: Could we use `Case::Sensitive` here? It feels a bit odd to have a catch-all.

**Review by jszwedko — APPROVED** [review]
Nice, makes sense to me! Does this introduce the requirement that the CSV text columns are UTF-8? Or did we already have that requirement?

**Review by spencergilbert — COMMENTED** [review]
Agree with both @jszwedko's comments

File: lib/enrichment/src/find_enrichment_table_records.rs (hunk 1 above)
Comment (StephenWakely) [reply lib/enrichment/src/find_enrichment_table_records.rs:128]: Strange. That change should not be there.

I think in theory we should be able to make this function infallible, but it's not quite there yet.

File: src/enrichment_tables/file.rs (hunk 2 above)
Comment (StephenWakely) [reply src/enrichment_tables/file.rs:195]: The pattern is also checking that the other two parameters are `Value::Bytes`. So the catch-all means it doesn't matter what the `Case` is when we don't have `Value::Bytes`.

Comment (jszwedko) [reply src/enrichment_tables/file.rs:195]: Ah, true, I missed that check for `Value::Bytes`. 👍

**Review by spencergilbert — APPROVED** [review]

**Comment by StephenWakely** [comment]
> Does this introduce the requirement that the CSV text columns are UTF-8? Or did we already have that requirement?

We have always required UTF-8 because the CSV parser returns String.

Other than that the only place we check for UTF-8 is if you wish to do a case insensitive search. If the column isn't valid UTF-8 then it falls back to comparing the bytes directly.

That requirement hasn't changed since it has been doing case insensitive searches from the start. This PR allows you to specify case sensitive which bypasses all UTF-8 validation.
